# Программирование на Python 

# Цикл `for`. Списки и кортежи. Списковые включения. Функция `map()`.


*Автор: Паршина Анастасия, НИУ ВШЭ (tg: @aaparshina)*

## Содержание


1. [Типы данных — часть 3. Списки и кортежи](#par1)
2. [Цикл `for` и функция `range()`](#par2)
3. [Списковые включения](#par3)
4. [И напоследок про `map()`](#par4)
4. [Дополнительные материалы](#parlast)

## Типы данных — часть 3. Списки и кортежи <a name="par1"></a>

Ранее мы уже обсуждали списки и кортежи в разговоре о типах данных. Тогда основное отличие заключалось в том, что списки — это изменяемый тип данных, а кортежи — неизменяемый. Проверить, что кортежи нельзя изменить, очень просто: 

In [33]:
(1, 2, 3)[0] # индексация работает 

1

In [34]:
(1, 2, 3)[0] = 0 # а вот попытка изменить кортеж (заменить его элемент) выдает ошибку

TypeError: 'tuple' object does not support item assignment

Оба типа данных представляют собой последовательности, однако кортеж — это неизменяемый и более быстрый аналог списка. Он защищает хранимые данные от непреднамеренных изменений и может использоваться в качестве ключа в словарях. 

In [32]:
import timeit
print(timeit.timeit('x=(1,2,3,4,5,6,7,8,9)', number=100000))
print(timeit.timeit('x=[1,2,3,4,5,6,7,8,9]', number=100000))

0.007532375000664615
0.01270995800041419


Также, если функция возвращает несколько значений, то делает она это в виде кортежа. 

In [35]:
import math 
math.modf(2.34) # к слову, как вы думаете, что делает эта функция?

(0.33999999999999986, 2.0)

Методы `.index()`, `.count()` и функция `len()` работают как с кортежами, так и со списками. Также списки можно объединять, НО это никак не меняет их, а буквально создает новый объект. 

In [39]:
A = (1, 2, 4) 
B = (1, 4)

print(A, id(A))
print(B, id(B))

C = A + B
print(A, id(A))
print(B, id(B))
print(C, id(C))

(1, 2, 4) 140561154781120
(1, 4) 140560329080576
(1, 2, 4) 140561154781120
(1, 4) 140560329080576
(1, 2, 4, 1, 4) 140560329040336


Со списками работает больше методов, все их обсудим в следующий раз. Пока остановимся только на двух, которые на данный момент могут представлять интерес, они оба имеют оношение к спискам, однако являются методами строк. Речь идет про `.join()` и `split()`. 



Метод `.join()` из списка <b>строк</b> делает строку. 

In [45]:
', '.join(['1234', '1246', '1748'])

'1234, 1246, 1748'

Фактически метод применяется к строке `', '` и говорит — используй эту строку в качестве разделителя между другими строками нашего списка.

In [48]:
', '.join(['1234', 1246, '1748'])

TypeError: sequence item 1: expected str instance, int found

Если же мы попытаемся объединить так список, состоящий не только из строк, то программа выдаст уже известную нам ошибку.

У этого метода есть еще пара нюансов, которые стоит запомнить. Помните про форматировани строки и f-строки? Попробуем подставить в такую запись `.join()`.

In [49]:
print(f'В магазине надо купить: {', '.join(["яблоки", "бананы", "апельсины"])}')

SyntaxError: f-string: expecting '}' (1763782797.py, line 1)

Это тот случай, где кавычки важны! Обратите внимание ошибка возникла потому, что f-строка подумала, что она заканчивается вот здесь `В магазине надо купить: {'` и при этом не нашла закрывающуюся скобочку `}`. Как решить проблему? Например, использовать в `.join()` двойные кавычки.

In [50]:
print(f'В магазине надо купить: {", ".join(["яблоки", "бананы", "апельсины"])}')

В магазине надо купить: яблоки, бананы, апельсины


И второй нюанс в том, что визуально может быть похоже следующее: 

In [51]:
print(", ".join(["яблоки", "бананы", "апельсины"]))
print(*["яблоки", "бананы", "апельсины"], sep = ', ')

яблоки, бананы, апельсины
яблоки, бананы, апельсины


В первом случае у нас создается строка, во втором — мы просто распаковываем имеющийся список.

Метод `split()` работает в обратную сторону — он буквально разделит нашу строку, и по умолчанию сделает это по пробелам.

In [60]:
years = input()
years = years.split()
print(years)

1258     1638 1729 982
['1258', '1638', '1729', '982']


Во-первых, обращаем внимание, что убираются все пробелы между символами. 

Во-вторых, это все можно было записать в одну строку вот так:

In [57]:
print(input().split()) # только тут у нас не создается переменная

1258     1638 1729 982
['1258', '1638', '1729', '982']


In [58]:
print(input().split(', '))

1258     1638 1729 982
['1258     1638 1729 982']


Если тоже самое «разделить» по запятой с пробелом, то получится список, состоящий из одной строки. Однако, если бы в ней были `, `, то наш метод также бы разделил по ним строку. 

Вернемся к нашему списку `years` и попробуем вытащить оттуда максимальное значение:

In [59]:
max(years)

'982'

Ой, что-то пошло не так. Это связано с тем, что было вытащено максимальное/первое значение по алфавиту, ведь мы работаем со строками. Нужно попробовать переделать наш список в список целых чисел. Быстрее всего это сделать с помощью функции `map()`, однако сначала рассмотрим, как будет работать цикл `for`.

## Цикл `for` и функция `range()` <a name="par2"></a>

Цикл `for` буквально можно воспринимать следующим обращом `для каждого элемента последовательности сделай что-то`. Под последовательностью здесь понимается <b>итерируемый объект</b> т.е. буквально тот, который вы можете растащить на части (например, строку можно ратащить на символы, список — на элементы списка, кортеж — на элементы кортежа, словарь — на пары ключ-значение). Небольшоше уточнение — разница, между последовательностями и итерируемымыи объектами, заключается в том, что в последовательностях элементы упорядочены.

Попробуем перебрать все элементы нашего списка `years` и напечатать каждый.

In [61]:
for y in years:
    print(y)

1258
1638
1729
982


Разберемеся с тем, что только что произошло. 

Во-первых, у нас прошло четыре <b>итерации</b> цикла, то есть четыре повторения цикла (это слово я еще при обсуждении цикла `while` вам говорила. Воспринимается это так: 

<i>Итерация №1</i>

    y = '1258'
    print(y) # 1258
    
<i>Итерация №2</i>

    y = '1638'
    print(y) # 1638
    
<i>Итерация №3</i>

    y = '1729'
    print(y) # 1729
    
<i>Итерация №4</i>

    y = '982'
    print(y) # 982
    
Во-вторых, подумайте, что сейчас хранится в переменной `y`?    

In [62]:
print(y)

982


В нашей переменной сохранилось именно последнее значение, которое она приняла в цикле. Фактически нам это не особо важно: сама переменная важна лишь внутри самого цикла, она там создается, потом с ней что-то происходит (в нашем случае мы ее печатаем), и затем мы о ней забываем. 

Доказать, что это буквально новая переменная, а не элемент списка, можно следующим образом.  

In [63]:
print(years)

for y in years:
    y = y + '1'
    print(y)
    
print(years)

['1258', '1638', '1729', '982']
12581
16381
17291
9821
['1258', '1638', '1729', '982']


Обратите внимание, что элементы списка не изменились. В цикле менялась только переменная `y`, там же она печаталась, а потом принимала новое значение. 

В-третьих, как цикл понимает, что пора остановиться? Ведь мы не прописываем никакое условие и не прописываем `break`. Сейчас будет сложно. 

У итерируемых объектов есть итератор. Мы можем сказать, что итератор — это объект, позволяющий реализовать поток данных и предлагающий возможность двигаться по нему. 

In [72]:
years_iterator = iter(years)
print(years_iterator)

Для реализации протокола итератора изпользуется встроенная функция `next()`. По сути, это инструмент для прохода по потоку данных, в нашем случае — по всем элементам списка. 

Когда мы дойдем до конца, она вызовет специальное исключение `StopIteration`, сигнализирующее о том, что итератор исчерпал все доступные ему значения.

In [73]:
print(next(years_iterator)) # первый элемент списка 
print(next(years_iterator)) # второй элемент списка
print(next(years_iterator)) # третий элемент списка
print(next(years_iterator)) # последний элемент списка
print(next(years_iterator)) # упс, элемент списка кончились

1258
1638
1729
982


StopIteration: 

Собственно, цикл `for` умеет ловить это исключение и останавливаться. 

По сути, цикл `for` — красивая «обертка» вот этого: 

In [75]:
years_iterator = iter(years)
running = True

while running:
    try:
        year = next(years_iterator)
        print(year)
    except StopIteration:
        running = False

1258
1638
1729
982


Пара слов о конструкции `try-except`: буквально «попытайся сделать так, а если не получится, то сделай эдак». Проще всего объяснить на пример: давайте поделим 10 на 0.

In [76]:
print(10/0)

ZeroDivisionError: division by zero

Да-да мы помним, что делить на 0 нельзя, об этом и Python нам напоминает. Возникает `ZeroDivisionError`. А теперь попробуем ее избежать с помощью `try-except`.

In [79]:
try:
    print(float('2,4'))
except:
    print('Ой, какая-то ошибка')

Ой, какая-то ошибка


Заметьте, просто `except` ловит вообще все ошибки, и лучше его не использовать (так мы рискуем пропустить какую-то важную ошибку, о которой хотелось бы знать). Лучше сразу указывать, что именно мы ловим: 

In [83]:
try:
    print(float('2,4'))
except ZeroDivisionError:
    print('Ой, на ноль делить нельзя')

ValueError: could not convert string to float: '2,4'

Код выше поймает только `ZeroDivisionError`, а об остальных нам скажет.

In [82]:
try:
    print(10/0)
except ZeroDivisionError:
    print('Ой, на ноль делить нельзя')

Ой, на ноль делить нельзя


Теперь, когда мы, вроде как, поняли, что лежит внутри цикла `for`, подумаем, а как в итоге заменить элементы списка. 

В этом нам поможет индексация. Как создать набор индексов? Тут есть два способа: 

1. функция `range()`
2. функция `enumerate()`

Первая универсальнее, так как буквально создает диапазон значений от и до (даже может сделать его с определенным шагом). При работе с ней нужно запомнить следующее: 

   `range(START, STOP, STEP)`
   
   + `START` — начало нашего диапазона
   + `STOP` — конец нашего диапазона (не включается!)
   + `STEP` — шаг, с которым мы идем 

In [86]:
# например, сделаем диапазон целых чисел от 1 до 10 с шагом 2 
# то есть возьем все нечетные числа от 1 до 10
for i in range(1, 11, 2):
    print(i)

1
3
5
7
9


In [87]:
# значения, которые можно подставить вместо «индексов» можно создать так: 
for i in range(len(years)):
    print(i)

0
1
2
3


В примере выше важно следующее: 
+ по умолчанию в значение `START` подставляется `0`;
+ как раз тут важно, что последнее значение (`4`) не включается — если бы включалось, то подставив его вместо индекса в список `years`, мы бы получили ошибку `IndexError`: 

In [88]:
print(years[4])

IndexError: list index out of range

Функция `enumerate()` вернет нам пару индекс-элемент, впредставленную в виде кортежа. 

In [91]:
for i in enumerate(years):
    print(i)
    print(f'Индексу {i[0]} соответствует элемент {i[1]}') # так можно разбить эту пару

(0, '1258')
Индексу 0 соответствует элемент 1258
(1, '1638')
Индексу 1 соответствует элемент 1638
(2, '1729')
Индексу 2 соответствует элемент 1729
(3, '982')
Индексу 3 соответствует элемент 982


Вернемся к идее поменять строки на целые числа в нашем списке `years`.

In [92]:
print(years)

for i in range(len(years)):
    years[i] = int(years[i])

print(years)

['1258', '1638', '1729', '982']
[1258, 1638, 1729, 982]


В цикле у нас создается переменная `i`, однако она используется для обращения к объектам списка, которые мы непосредственно изменяем. 



## Списковые включения <a name="par3"></a>

Если мы не хотим изменять текущий список, а хотим сохранить значения в новый, то это можно сделать так: 

In [93]:
years = input().split()

int_years = []

for i in range(len(years)):
    int_years.append(int(years[i]))
    
print(years)
print(int_years)

1258     1638 1729 982
['1258', '1638', '1729', '982']
[1258, 1638, 1729, 982]


Да, уже выглядит сложнее. Здесь мы создаем пустой список, в который складываем наши новые объекты. Метод `.append()` мы уже встречали (он в конец списка добавляет новые значения). Более подробно о методах поговорим в следующий раз!

Код безумно длинный (да-да!). Его можно записать в одну строчку, используя списковые включения.

In [94]:
int_years_2 = [int(years[i]) for i in range(len(years))]
print(int_years_2)

[1258, 1638, 1729, 982]


Мы буквально создаем пустой список, говорим, положи туда `int(years[i])`, а потом указываем, откуда это `i` берется.

Списковые включения достаточно короткая конструкция. Максимум — можно добавить в него условие (но без `elif`). 

In [99]:
# вот так добавляется просто if
even_numbers = [i for i in range(1, 11) if i%2 == 0]
print(even_numbers)

[2, 4, 6, 8, 10]


In [100]:
# а вот так if-else 
even_numbers = [i if i%2 == 0 else f'{i} - нечетное' for i in range(1, 11)]
print(even_numbers)

['1 - нечетное', 2, '3 - нечетное', 4, '5 - нечетное', 6, '7 - нечетное', 8, '9 - нечетное', 10]


Конструкцию с `elif` вы туда не запишете.

Сделать списковое включение в списковом включении, конечно, можно... главное, не запутайтесь :) 

In [98]:
[i**2 for i in [i**2 for i in range(10)]]

[0, 1, 16, 81, 256, 625, 1296, 2401, 4096, 6561]

## И напоследок про `map()` <a name="par4"></a>

Вспомним про наш список `years`. Преобразовать все в целые числа можно еще и так: 

In [101]:
int_years_3 = list(map(int, years))
print(years)
print(int_years_3)

['1258', '1638', '1729', '982']
[1258, 1638, 1729, 982]


Что тут произошло? Функция `map()` буквально говорит: возьми функцию `int()` и примени ее ко всем элементам нашего списка. Возвращает нам объект типа `map`, который потом мы уже преобразовываем в список с помощью функции `list()`.

Функция `map()` может работать не только с встроенными функциями, но и с любыми другими, даже написанными именно нами.

In [105]:
from math import sqrt

sqrt_years = list(map(sqrt, int_years_3))

print(int_years_3)
print(sqrt_years)

[1258, 1638, 1729, 982]
[35.4682957019364, 40.47221268969612, 41.58124577258358, 31.336879231984796]


Например, возьмем квадратный корень из всех наших значений. 

В заключении скажем, что есть еще нечто, называемое безымянными функциями `lambda`, которые также работают с `map()`, но о них мы поговорим чуть позже, при обсуждении функций. Не переживаем, если все, что написано ниже, буде непонятно!! Мы разберемся со всем!

In [106]:
plus_one_years = list(map(lambda x: x + 1, int_years_3))

print(int_years_3)
print(plus_one_years)

[1258, 1638, 1729, 982]
[1259, 1639, 1730, 983]


А и началось же все, что мы хотели максимальный год достать. Вот он: 

In [107]:
print(max(int_years))

1729


## Дополнительные материалы <a name="parlast"></a>

+ Документация Python [Built-in Types](https://docs.python.org/3/library/stdtypes.html)
+ Документация Python [More Control Flow Tools](https://docs.python.org/3/tutorial/controlflow.html)
+ Щуров И.В., Тамбовцева А.А., Жучкова С.В. —  курс «Основы программирования в Python» ([ссылка на курс](https://allatambov.github.io/pypolit/pypolit.html))
+ Статья на Хабр. [Понимание итераторов в Python](https://habr.com/ru/post/488112/)
+ [Цикл `for` в Python - как работает, синтаксис, примеры](https://pythonchik.ru/osnovy/cikl-for-v-python)
+ [Как работает функция `enumerate()` в Python?](https://pythonist.ru/kak-rabotaet-funkcziya-enumerate-v-python/)
+ Python для начинающих. [Про `try-except`](https://pythonworld.ru/tipy-dannyx-v-python/isklyucheniya-v-python-konstrukciya-try-except-dlya-obrabotki-isklyuchenij.html)